Integration of error functions and trigonometric/hyperbolic functions combinations

salastro opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Solves some more integrals in #27586. 

#### Brief description of what is fixed or changed

Added a new rewriting strategy `trig_cmplx_exp_rule` to `manualintegrate.py`. This rule rewrites `sin`, `cos`, `sinh`, and `cosh` in terms of complex exponentials when the integrand contains `erf`, `erfc`, `erfi`. This solves problems of the form `integrate erf(x)*sin(x)` as they cannot be solved in any other implemented way. It is easy to solve it when turned into complex exponentials as erf rule is already implemented.
The reason error functions are a condition is to avoid messing with other integrals where cyclic integration happens, they still give correct results but they fail tests because of the different answer form.


#### Other comments

It will be better if there is a way to simplify the final answer back into trigonometric/hyperbolic functions instead of the ugly complex exponentials form. When I try to use `rewrite` it adds `cosh` and `sinh` terms to the expression which is even uglier.

```
In [2]: integrate(erf(x)*sin(x), manual=True)
Out[2]: 
   ⎛     ⅈ⋅x             -1/4    ⎛    ⅈ⎞      -1/4    ⎛    ⅈ⎞      -ⅈ⋅x       ⎞ 
-ⅈ⋅⎜- ⅈ⋅ℯ   ⋅erf(x) + ⅈ⋅ℯ    ⋅erf⎜x - ─⎟ + ⅈ⋅ℯ    ⋅erf⎜x + ─⎟ - ⅈ⋅ℯ    ⋅erf(x)⎟ 
   ⎝                             ⎝    2⎠              ⎝    2⎠                 ⎠ 
────────────────────────────────────────────────────────────────────────────────
                                       2                                        

In [3]: integrate(erf(x)*sin(x), manual=True).simplify()
Out[3]: 
                 -1/4    ⎛    ⅈ⎞    -1/4    ⎛    ⅈ⎞               
   ⅈ⋅x          ℯ    ⋅erf⎜x - ─⎟   ℯ    ⋅erf⎜x + ─⎟    -ⅈ⋅x       
  ℯ   ⋅erf(x)            ⎝    2⎠            ⎝    2⎠   ℯ    ⋅erf(x)
- ─────────── + ──────────────── + ──────────────── - ────────────
       2               2                  2                2      

In [4]: integrate(erf(x)*sin(x), manual=True).simplify().rewrite(cos)
Out[4]: 
                                                                                         ⎛    ⅈ⎞                               ⎛    ⅈ⎞
                                                             (-sinh(1/4) + cosh(1/4))⋅erf⎜x - ─⎟   (-sinh(1/4) + cosh(1/4))⋅erf⎜x + ─⎟
  (-ⅈ⋅sin(x) + cos(x))⋅erf(x)   (ⅈ⋅sin(x) + cos(x))⋅erf(x)                               ⎝    2⎠                               ⎝    2⎠
- ─────────────────────────── - ────────────────────────── + ─────────────────────────────────── + ───────────────────────────────────
               2                            2                                 2                                     2                 

``` 

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* integrals
  * Added support for combinations of error functions and trigonometric/hyperbolic functions.
<!-- END RELEASE NOTES -->

salastro committed:
manualintegrate: add a rule to rewrite trig/hyperbolic functions as exponentials when erf functions present

oscarbenjamin commented on sympy/integrals/manualintegrate.py:
```diff
@@ -1424,6 +1424,35 @@ def trig_product_rule(integral: IntegralInfo):
+    # Replace trig and hyperbolic functions with their exponential forms
+    replaced = integrand.replace(
+        lambda expr: expr.func in (sin, cos, sinh, cosh),
+        lambda expr: expr.rewrite(exp)
```
You can do this directly with `rewrite`:
```python
In [6]: (sin(x) + cos(x)).rewrite([sin, exp], tan)
Out[6]: 
               ⎛x⎞  
          2⋅tan⎜─⎟  
               ⎝2⎠  
cos(x) + ───────────
            2⎛x⎞    
         tan ⎜─⎟ + 1
             ⎝2⎠ 
```
It is probably better though to only do this for the expressions that contain the integration variable.

salastro commented:
> > In [2]: integrate(erf(x)*sin(x), manual=True)
> > Out[2]:
> > ⎛     ⅈ⋅x             -1/4    ⎛    ⅈ⎞      -1/4    ⎛    ⅈ⎞      -ⅈ⋅x       ⎞
> > -ⅈ⋅⎜- ⅈ⋅ℯ   ⋅erf(x) + ⅈ⋅ℯ    ⋅erf⎜x - ─⎟ + ⅈ⋅ℯ    ⋅erf⎜x + ─⎟ - ⅈ⋅ℯ    ⋅erf(x)⎟
> > ⎝                             ⎝    2⎠              ⎝    2⎠                 ⎠
> > ────────────────────────────────────────────────────────────────────────────────
> > 2
> 
> I think this is a bad idea. Starting from real expressions should not go to complex expressions. For `sinh` and `cosh` this is not so bad but I would still prefer to keep `sinh` and `cosh` rather than convert them to `exp`.

The answer is a complex expression either way:
<img width="793" height="826" alt="image" src="https://github.com/user-attachments/assets/4d670ddd-e811-494b-b9c4-f79ec2bd0328" />

Ng, E. W., & Geller, M. (1969). A table of integrals of the error functions. Journal of Research of the National Bureau of Standards, Section B: Mathematical Sciences, 73B(1), 1–20. https://doi.org/10.6028/jres.073B.001

oscarbenjamin commented:
> The answer is a complex expression either way:

If it is unavoidable to use some complex expressions that that can be done but if the trigs are real then it is better to keep them. This should be the expected output:
```python
In [32]: e = e.rewrite(exp, cos).expand().rewrite([sinh, cosh], exp).collect(e.atoms(erf))

In [33]: e.collect(e.atoms(exp), factor_terms)
Out[33]: 
⎛   ⎛    ⅈ⎞      ⎛    ⅈ⎞⎞  -1/4                
⎜erf⎜x - ─⎟ + erf⎜x + ─⎟⎟⋅ℯ                    
⎝   ⎝    2⎠      ⎝    2⎠⎠                      
─────────────────────────────── - cos(x)⋅erf(x)
               2 
```
Only the part in brackets involves complex numbers but it is real valued for real `x`.

salastro commented:
> ```python
> collect(e.atoms(exp), factor_terms)
> ```

Is it acceptable to do these simplification steps in the `manualintegrate` itself? So something like
```python
def manualintegrate(f, var):
    # ...
    result = integral_steps(f, var).eval()
    if (has_cmplx_exp_cond):
            result = result.rewrite(exp, cos).expand().rewrite([sinh, cosh], exp).collect(result.atoms(erf)).collect(result.atoms(exp), factor_terms)
   # ...
   return result
```

Or is there another approach?

salastro committed:
manualintegrate: detect and simplify symmetric complex exponentials before returning result

salastro commented:
With this I get:
```
In [1]: integrate(erf(x)*sin(x), manual=True)
Out[1]:
⎛   ⎛    ⅈ⎞      ⎛    ⅈ⎞⎞  -1/4
⎜erf⎜x - ─⎟ + erf⎜x + ─⎟⎟⋅ℯ
⎝   ⎝    2⎠      ⎝    2⎠⎠
─────────────────────────────── - cos(x)⋅erf(x)
               2
In [2]: integrate(erf(x)*cos(x), manual=True)
Out[2]:
  ⎛   ⎛    ⅈ⎞      ⎛    ⅈ⎞⎞  -1/4
ⅈ⋅⎜erf⎜x - ─⎟ - erf⎜x + ─⎟⎟⋅ℯ
  ⎝   ⎝    2⎠      ⎝    2⎠⎠
───────────────────────────────── + sin(x)⋅erf(x)
                2
In [3]:
```

oscarbenjamin commented on sympy/integrals/manualintegrate.py:
```diff
@@ -2003,6 +2030,42 @@ def rewrites_rule(integral):
+
+    for sub in preorder_traversal(expr):
+        if sub.func == exp:
+            arg = simplify(sub.args[0])
```
The `simplify` function should not be used in library code.

salastro replied:
Why? It is already used all over the manualintegrate module.
What alternative should I use? I am using it to normalize expressions to check if they are the same or not. So `exp(-x)+exp(-(-x))` is detected as a valid match.

oscarbenjamin replied:
The simplify function is slow and heuristic. It leads to unpredictable outputs and it slows everything down.

If there is a need for some simplification in library code then you should figure out which specific simplification function is needed rather than calling simplify which heuristically tries many of them.

salastro replied:
Makes sense. I will apply some steps like expand, factor terms, powsimp; and see where that is going.

salastro committed:
manualintegrate: simplify symmetric exponentials into hyperbolic functions before returning result

salastro committed:
manualintegrate: avoid using `simplify` in `has_cmplx_exp_pair`

salastro commented:
This should be better.

salastro commented:
I do not know how or why does `sqrt(a)` gets rewritten as `exp(1/(18*log(x)))` while they are clearly not equal.

oscarbenjamin commented:
> I do not know how or why does `sqrt(a)` gets rewritten as `exp(1/(18*log(x)))` while they are clearly not equal.

Can you demonstrate what you mean by showing the code that does that?

salastro commented:
The thing is we don't get `exp(logx/2)` which is correct. We get `exp(1/18*logx)` which is incorrect.

oscarbenjamin commented:
> We get `exp(1/18*logx)`

What is the expression that is rewritten to this?

```python
In [1]: e = x**(S(1)/18)

In [2]: e
Out[2]: 
18___
╲╱ x 

In [3]: e.rewrite(exp)
Out[3]: 
 log(x)
 ──────
   18  
ℯ 
```

salastro commented:
> Looking at the failed test case the result is looking pretty ugly
> 
> ```
> In [12]: eqs4 = [Eq(Derivative(f(x), x), x*(f(x) + g(x) + h(x)) + sin(x)),
>     ...:             Eq(Derivative(g(x), x), x*(f(x) + g(x) + h(x)) + sin(x)),
>     ...:             Eq(Derivative(h(x), x), x*(f(x) + g(x) + h(x)) + sin(x))]
> 
> In [13]: dsolve(eqs4)
> Out[13]: 
> ⎡                                                                                2                          ⎛       1              1    ⎞                                  2                  ↪
> ⎢                                                                             3⋅x       1          1        ⎜   ─────────      ─────────⎟                               3⋅x       1           ↪
> ⎢                                             2                               ────  ─────────  ─────────    ⎜   18⋅log(6)      18⋅log(6)⎟                               ────  ─────────  ──── ↪
> ⎢                                          3⋅x                                 2    18⋅log(6)  18⋅log(π)    ⎜x⋅ℯ            ⅈ⋅ℯ         ⎟                                2    18⋅log(6)  18⋅l ↪
> ⎢                                          ────   ⅈ⋅(-cosh(1/6) + sinh(1/6))⋅ℯ    ⋅ℯ         ⋅ℯ         ⋅erf⎜──────────── - ────────────⎟   ⅈ⋅(-cosh(1/6) + sinh(1/6))⋅ℯ    ⋅ℯ         ⋅ℯ     ↪
> ⎢         C₁   C₂   2⋅C₃   ⎛C₁   C₂   C₃⎞   2                                                               ⎝     2              6      ⎠                                                     ↪
> ⎢f(x) = - ── - ── + ──── + ⎜── + ── + ──⎟⋅ℯ     + ─────────────────────────────────────────────────────────────────────────────────────── - ───────────────────────────────────────────────── ↪
> ⎣         3    3     3     ⎝3    3    3 ⎠                                                   12                                                                                        12      ↪
> 
> ↪          ⎛       1              1    ⎞                                                                                2                          ⎛       1              1    ⎞              ↪
> ↪ 1        ⎜   ─────────      ─────────⎟                                                                             3⋅x       1          1        ⎜   ─────────      ─────────⎟              ↪
> ↪ ─────    ⎜   18⋅log(6)      18⋅log(6)⎟                                             2                               ────  ─────────  ─────────    ⎜   18⋅log(6)      18⋅log(6)⎟              ↪
> ↪ og(π)    ⎜x⋅ℯ            ⅈ⋅ℯ         ⎟                                          3⋅x                                 2    18⋅log(6)  18⋅log(π)    ⎜x⋅ℯ            ⅈ⋅ℯ         ⎟              ↪
> ↪      ⋅erf⎜──────────── + ────────────⎟                                          ────   ⅈ⋅(-cosh(1/6) + sinh(1/6))⋅ℯ    ⋅ℯ         ⋅ℯ         ⋅erf⎜──────────── - ────────────⎟   ⅈ⋅(-cosh(1 ↪
> ↪          ⎝     2              6      ⎠         2⋅C₁   C₂   C₃   ⎛C₁   C₂   C₃⎞   2                                                               ⎝     2              6      ⎠              ↪
> ↪ ──────────────────────────────────────, g(x) = ──── - ── - ── + ⎜── + ── + ──⎟⋅ℯ     + ─────────────────────────────────────────────────────────────────────────────────────── - ────────── ↪
> ↪                                                 3     3    3    ⎝3    3    3 ⎠                                                   12                                                         ↪
> 
> ↪                      2                          ⎛       1              1    ⎞                                                                                  2                          ⎛ ↪
> ↪                   3⋅x       1          1        ⎜   ─────────      ─────────⎟                                                                               3⋅x       1          1        ⎜ ↪
> ↪                   ────  ─────────  ─────────    ⎜   18⋅log(6)      18⋅log(6)⎟                                               2                               ────  ─────────  ─────────    ⎜ ↪
> ↪                    2    18⋅log(6)  18⋅log(π)    ⎜x⋅ℯ            ⅈ⋅ℯ         ⎟                                            3⋅x                                 2    18⋅log(6)  18⋅log(π)    ⎜ ↪
> ↪ /6) + sinh(1/6))⋅ℯ    ⋅ℯ         ⋅ℯ         ⋅erf⎜──────────── + ────────────⎟                                            ────   ⅈ⋅(-cosh(1/6) + sinh(1/6))⋅ℯ    ⋅ℯ         ⋅ℯ         ⋅erf⎜ ↪
> ↪                                                 ⎝     2              6      ⎠           C₁   2⋅C₂   C₃   ⎛C₁   C₂   C₃⎞   2                                                               ⎝ ↪
> ↪ ─────────────────────────────────────────────────────────────────────────────, h(x) = - ── + ──── - ── + ⎜── + ── + ──⎟⋅ℯ     + ─────────────────────────────────────────────────────────── ↪
> ↪                                 12                                                      3     3     3    ⎝3    3    3 ⎠                                                   12                ↪
> 
> ↪        1              1    ⎞                                  2                          ⎛       1              1    ⎞⎤
> ↪    ─────────      ─────────⎟                               3⋅x       1          1        ⎜   ─────────      ─────────⎟⎥
> ↪    18⋅log(6)      18⋅log(6)⎟                               ────  ─────────  ─────────    ⎜   18⋅log(6)      18⋅log(6)⎟⎥
> ↪ x⋅ℯ            ⅈ⋅ℯ         ⎟                                2    18⋅log(6)  18⋅log(π)    ⎜x⋅ℯ            ⅈ⋅ℯ         ⎟⎥
> ↪ ──────────── - ────────────⎟   ⅈ⋅(-cosh(1/6) + sinh(1/6))⋅ℯ    ⋅ℯ         ⋅ℯ         ⋅erf⎜──────────── + ────────────⎟⎥
> ↪      2              6      ⎠                                                             ⎝     2              6      ⎠⎥
> ↪ ──────────────────────────── - ───────────────────────────────────────────────────────────────────────────────────────⎥
> ↪                                                                          12                                           ⎦
> ```

If you look at this, you will find, for example, `exp(1/18*log(6))` which is incorrect as the final answer has `sqrt(6)` instead. As you see:

```
In [1]: integrate(exp(-x**2)*sin(x), manual=True)
Out[1]: 
                                        1    
                                     ────────
  ⎛     ⎛    ⅈ⎞      ⎛    ⅈ⎞⎞  -1/4  8⋅log(π)
ⅈ⋅⎜- erf⎜x - ─⎟ + erf⎜x + ─⎟⎟⋅ℯ    ⋅ℯ        
  ⎝     ⎝    2⎠      ⎝    2⎠⎠                
─────────────────────────────────────────────
                      4                      

In [2]: 
```

From Wolfram Alpha:

<img width="1071" height="199" alt="image" src="https://github.com/user-attachments/assets/30f2b98d-df37-4404-b1a8-b50223fbe75f" />

oscarbenjamin commented:
I suggest using a debugger to see where something invalid happens.

salastro committed:
manualintegrate: extend `trig_cmplx_exp_rule` to match Gaussian exponentials

salastro committed:
manualintegrate: move `trig_cmplx_exp_rule` into Mul rules instead of generic placement

Avoids messing up other integrals where the rule is not needed/does not help.

salastro committed:
Replace unevaluated Integral in sysode test with explicit erf solution

salastro committed:
manualintegrate: guard trig_cmplx_exp_rule with exp check

This should have been added from the beginning because without it the
rule terminates on expressions it should rewrite.

This should mean "apply rewrite if found a guassian exponential, if
found a linear exponential, ignore it."

salastro committed:
manualintegrate: replace `rewrite_exp_hyperbolic` with builtin `exptrigsimp`

salastro commented:
I do not think there is a specific trig rule for this kind of integrals. The steps that I found uses this technique. Other method for integrating `erf(x)*sin(x)` would be doing IBP and then rewriting the `exp(-x**2)*cos(x)` integral as `ℜ{exp(-x**2)*exp(I*x)}` solving this integral part we get `ℜ{erf(x-I/2)}` so the final answer is `-erf(x)*cos(x)+exp(-1/4)*ℜ{erf(x-I/2)}`, which I believe is even harder to implement.

salastro commented:
There is already `exptrigsimp` which works just fine but cannot handle `exp(-I*x)/(2*I) - exp(I*x)/(2*I) + sin(x)` for some reason

```
In [14]: simplify(exp(-I*x)/(2*I)-exp(I*x)/(2*I)+sin(x))
Out[14]: 
   ⅈ⋅x               -ⅈ⋅x
ⅈ⋅ℯ               ⅈ⋅ℯ    
────── + sin(x) - ───────
  2                  2   

In [15]: simplify(exp(-I*x)/(2*I)-exp(I*x)/(2*I))
Out[15]: -sin(x)

In [16]: exptrigsimp(exp(-I*x)-exp(I*x))
Out[16]: -2⋅ⅈ⋅sin(x)

In [17]: exptrigsimp((exp(-I*x)-exp(I*x))/I)
Out[17]: -2⋅sin(x)

In [18]: exptrigsimp((exp(-I*x)-exp(I*x))+sin(x))
Out[18]: (1 - 2⋅ⅈ)⋅sin(x)

In [19]: exptrigsimp(exp(-I*x)/(2*I)-exp(I*x)/(2*I)+sin(x))
Out[19]: 
   ⅈ⋅x               -ⅈ⋅x
ⅈ⋅ℯ               ⅈ⋅ℯ    
────── + sin(x) - ───────
  2                  2   

In [20]: 
```

salastro commented:
That fails cause `cancel` cannot operate on other objects (e.g. Manifolds, Matrices, etc) so a different approach is needed.

oscarbenjamin commented:
> Other method for integrating `erf(x)*sin(x)` would be doing IBP and then rewriting the `exp(-x**2)*cos(x)`.

That looks like where the formulae would come from. I think it is better to use integration by parts and then to have something that handles exp/trig integrals which could be done with rewriting but it should be a very targeted rewrite-compute-restore.

salastro committed:
trigsimp: use `factor_terms` on Add nodes in `exp_trig` to expose cancellations

Helps with simplifying the case `exp(-I*x)/(2*I)-exp(I*x)/(2*I)+sin(x)`

Before:

```
In [1]: e = exp(-I*x)/(2*I)-exp(I*x)/(2*I)+sin(x)

In [2]: exptrigsimp(e)
Out[2]:
   ⅈ⋅x               -ⅈ⋅x
ⅈ⋅ℯ               ⅈ⋅ℯ
────── + sin(x) - ───────
  2                  2

In [3]:
```

After:
```
In [1]: e = exp(-I*x)/(2*I)-exp(I*x)/(2*I)+sin(x)

In [2]: exptrigsimp(e)
Out[2]: 0

In [3]:
```

salastro commented:
Weirdly enough, the failing test in `test_continuous_rv`

```
In [1]: from sympy.stats import cdf, Moyal
   ...: mu = Symbol('mu', real=True)
   ...: sigma = Symbol('sigma', positive=True)
   ...: M = Moyal('M', mu, sigma)
   ...: cdf(M)(z)
Out[1]: 
        z                    
        ⌠                    
        ⎮          -x   μ    
        ⎮          ───  ─    
        ⎮           σ   σ    
     μ  ⎮   -x   -ℯ   ⋅ℯ     
    ─── ⎮   ───  ─────────   
    2⋅σ ⎮   2⋅σ      2       
√2⋅ℯ   ⋅⎮  ℯ   ⋅ℯ          dx
        ⌡                    
        -∞                   
─────────────────────────────
           2⋅√π⋅σ            

In [2]: 
```

It should evaluate to

```
   ⎛                  ⎛     μ     z ⎞               ⎞     
   ⎜           -μ     ⎜    ─── - ───⎟            -μ ⎟   μ 
   ⎜           ───    ⎜    2⋅σ   2⋅σ⎟            ───⎟  ───
   ⎜           2⋅σ    ⎜√2⋅ℯ         ⎟            2⋅σ⎟  2⋅σ
√2⋅⎜- √2⋅√π⋅σ⋅ℯ   ⋅erf⎜─────────────⎟ + √2⋅√π⋅σ⋅ℯ   ⎟⋅ℯ   
   ⎝                  ⎝      2      ⎠               ⎠     
──────────────────────────────────────────────────────────
                          2⋅√π⋅σ                          
```

The issue is avoided after removing the simplification line in manualintegrate:

```python
result = result.rewrite(exp, cos).expand().rewrite([sinh, cosh], exp).collect(result.atoms(erf)).collect(result.atoms(exp), factor_terms)
```

How to debug this issue?

oscarbenjamin commented:
> > it should be a very targeted rewrite-compute-restore
> 
> I have no idea how to approach that.

I don't know the manualintegrate code so well but it has a bunch of rules for different integrals. The approach taken here is just to say that if the integrand contains any `exp(x**2)` and also has `erf` then try replacing all trig with exponentials which is quite a broad rule.

I think really what we want is something that says "to integrate `trig*erf` do integration by parts to get an integral with `trig*exp` and then have a special handler for `trig*exp` that knows how to convert trig to exp and back for the particular integrand."

salastro committed:
manualintegrate: narrow trig_exp_rule

- Restrict rule to integrands with Gaussian exponentials and trig/hyperbolic functions
- Remove post-processing of has_cmplx_exp_pair and exptrigsimp
- Remove unused imports i.e. expand, factor_terms, exptrigsimp,
  trigsimp, preorder_traversal

Error functions combinations are still handled by the IBP module. The
only downside is a bit more messy expressions.

salastro commented:
The code turns out to be much simpler than what I was doing; I had a mental block that hindered my progress significantly. Either way, this should work. The only downside is more messy expressions:

```
In [1]: integrate(sin(x)*erf(x))
Out[1]: 
  ⎛    -1/4    ⎛    ⅈ⎞       -1/4    ⎛    ⅈ⎞⎞                
  ⎜√π⋅ℯ    ⋅erf⎜x - ─⎟   √π⋅ℯ    ⋅erf⎜x + ─⎟⎟                
  ⎜            ⎝    2⎠               ⎝    2⎠⎟                
2⋅⎜─────────────────── + ───────────────────⎟                
  ⎝         4                     4         ⎠                
───────────────────────────────────────────── - cos(x)⋅erf(x)
                     √π                                      

In [2]: 
```

Instead of the previous more clean results due to the post-processing step.

salastro commented:
Now there is no need for any back-conversion as the combination `trig*exp2` always results in an erf/erfi pair.

oscarbenjamin commented:
This will need some tests.

salastro commented:
Should I add a condition to factor out these terms? It is really annoying having them around.

oscarbenjamin commented:
If it can be done in the right place then maybe factor_terms could be used.

oscarbenjamin commented:
No, no global flags.

salastro commented:
Is there a way to mark when terms need to get factored? My first idea was to make a global variable `_factor_terms_flag` like `_integral_cache`, `_parts_u_cache`, `_cache_dummy`, and other global variables in the codebase. Then mark it on when the `trig_cmplx_exp_rule` is called, and after that in `manualintegrate` have an if statement that executes `factor_terms` on the result before returning.

salastro committed:
Add tests for integrals of Gaussian exponentials with sin and cos factors

oscarbenjamin commented on sympy/integrals/tests/test_manual.py:
```diff
@@ -300,6 +300,10 @@ def test_manualintegrate_special():
     assert_is_integral_of(f, F)
     f, F = erfi(7*x)*exp(6*x), exp(6*x)*erfi(7*x)/6 - exp(-Rational(9,49))*erfi(7*x + Rational(3,7))/6
     assert_is_integral_of(f, F)
+    f, F = sin(2*x)*exp(-3*x**2), -I*(sqrt(3)*sqrt(pi)*exp(-Rational(1,3))*erf(sqrt(3)*(6*x - 2*I)/6)/6 - sqrt(3)*sqrt(pi)*exp(-Rational(1,3))*erf(sqrt(3)*(6*x + 2*I)/6)/6)/2
+    assert_is_integral_of(f, F)
+    f, F = cos(2*x)*exp(-3*x**2), sqrt(3)*sqrt(pi)*exp(-Rational(1,3))*erf(sqrt(3)*(6*x - 2*I)/6)/12 + sqrt(3)*sqrt(pi)*exp(-Rational(1,3))*erf(sqrt(3)*(6*x + 2*I)/6)/12
+    assert_is_integral_of(f, F)
```
I think there are many more cases that are enabled in this PR e.g.:
```
In [1]: integrate(sin(x)*erf(x), manual=True)
Out[1]: 
  ⎛    -1/4    ⎛    ⅈ⎞       -1/4    ⎛    ⅈ⎞⎞                
  ⎜√π⋅ℯ    ⋅erf⎜x - ─⎟   √π⋅ℯ    ⋅erf⎜x + ─⎟⎟                
  ⎜            ⎝    2⎠               ⎝    2⎠⎟                
2⋅⎜─────────────────── + ───────────────────⎟                
  ⎝         4                     4         ⎠                
───────────────────────────────────────────── - cos(x)⋅erf(x)
                     √π  
```

salastro replied:
I added
```python
    f, F = sin(x)*erf(x), 2*(sqrt(pi)*exp(-Rational(1,4))*erf(x - I/2)/4 + sqrt(pi)*exp(-Rational(1,4))*erf(x + I/2)/4)/sqrt(pi) - cos(x)*erf(x)
    assert_is_integral_of(f, F)
```
but the test fails because the lack of simplification
```
========================================================================================== FAILURES ===========================================================================================
________________________________________________________________________________ test_manualintegrate_special _________________________________________________________________________________

    def test_manualintegrate_special():
        f, F = 4*exp(-x**2/3), 2*sqrt(3)*sqrt(pi)*erf(sqrt(3)*x/3)
        assert_is_integral_of(f, F)
        f, F = 3*exp(4*x**2), 3*sqrt(pi)*erfi(2*x)/4
        assert_is_integral_of(f, F)
        f, F = x**Rational(1, 3)*exp(-x/8), -16*uppergamma(Rational(4, 3), x/8)
        assert_is_integral_of(f, F)
        f, F = exp(2*x)/x, Ei(2*x)
        assert_is_integral_of(f, F)
        f, F = exp(1 + 2*x - x**2), sqrt(pi)*exp(2)*erf(x - 1)/2
        assert_is_integral_of(f, F)
        f = sin(x**2 + 4*x + 1)
        F = (sqrt(2)*sqrt(pi)*(-sin(3)*fresnelc(sqrt(2)*(2*x + 4)/(2*sqrt(pi))) +
            cos(3)*fresnels(sqrt(2)*(2*x + 4)/(2*sqrt(pi))))/2)
        assert_is_integral_of(f, F)
        f, F = cos(4*x**2), sqrt(2)*sqrt(pi)*fresnelc(2*sqrt(2)*x/sqrt(pi))/4
        assert_is_integral_of(f, F)
        f, F = sin(3*x + 2)/x, sin(2)*Ci(3*x) + cos(2)*Si(3*x)
        assert_is_integral_of(f, F)
        f, F = sinh(3*x - 2)/x, -sinh(2)*Chi(3*x) + cosh(2)*Shi(3*x)
        assert_is_integral_of(f, F)
        f, F = 5*cos(2*x - 3)/x, 5*cos(3)*Ci(2*x) + 5*sin(3)*Si(2*x)
        assert_is_integral_of(f, F)
        f, F = cosh(x/2)/x, Chi(x/2)
        assert_is_integral_of(f, F)
        f, F = cos(x**2)/x, Ci(x**2)/2
        assert_is_integral_of(f, F)
        f, F = 1/log(2*x + 1), li(2*x + 1)/2
        assert_is_integral_of(f, F)
        f, F = polylog(2, 5*x)/x, polylog(3, 5*x)
        assert_is_integral_of(f, F)
        f, F = 5/sqrt(3 - 2*sin(x)**2), 5*sqrt(3)*elliptic_f(x, Rational(2, 3))/3
        assert_is_integral_of(f, F)
        f, F = sqrt(4 + 9*sin(x)**2), 2*elliptic_e(x, Rational(-9, 4))
        assert_is_integral_of(f, F)
        f, F = exp(-x**2)*exp(x), Rational(1,2)*exp(Rational(1,4))*sqrt(pi)*erf(x - Rational(1,2))
        assert_is_integral_of(f, F)
        f, F = erf(3*x)*exp(2*x), exp(2*x)*erf(3*x)/2 - exp(Rational(1,9))*erf(3*x - Rational(1,3))/2
        assert_is_integral_of(f, F)
        f, F = erfc(4*x)*exp(5*x), exp(5*x)*erfc(4*x)/5 + exp(Rational(25,64))*erf(4*x - Rational(5,8))/5
        assert_is_integral_of(f, F)
        f, F = erfi(7*x)*exp(6*x), exp(6*x)*erfi(7*x)/6 - exp(-Rational(9,49))*erfi(7*x + Rational(3,7))/6
        assert_is_integral_of(f, F)
        f, F = sin(2*x)*exp(-3*x**2), -I*(sqrt(3)*sqrt(pi)*exp(-Rational(1,3))*erf(sqrt(3)*(6*x - 2*I)/6)/6 - sqrt(3)*sqrt(pi)*exp(-Rational(1,3))*erf(sqrt(3)*(6*x + 2*I)/6)/6)/2
        assert_is_integral_of(f, F)
        f, F = cos(2*x)*exp(-3*x**2), sqrt(3)*sqrt(pi)*exp(-Rational(1,3))*erf(sqrt(3)*(6*x - 2*I)/6)/12 + sqrt(3)*sqrt(pi)*exp(-Rational(1,3))*erf(sqrt(3)*(6*x + 2*I)/6)/12
        assert_is_integral_of(f, F)
        f, F = sin(x)*erf(x), 2*(sqrt(pi)*exp(-Rational(1,4))*erf(x - I/2)/4 + sqrt(pi)*exp(-Rational(1,4))*erf(x + I/2)/4)/sqrt(pi) - cos(x)*erf(x)
>       assert_is_integral_of(f, F)

sympy/integrals/tests/test_manual.py:308: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

f = sin(x)*erf(x), F = (sqrt(pi)*exp(-1/4)*erf(x - I/2)/2 + sqrt(pi)*exp(-1/4)*erf(x + I/2)/2)/sqrt(pi) - cos(x)*erf(x)

    def assert_is_integral_of(f: Expr, F: Expr):
>       assert manualintegrate(f, x) == F
E       assert 2*(sqrt(pi)*exp(-1/4)*erf(x - I/2)/4 + sqrt(pi)*exp(-1/4)*erf(x + I/2)/4)/sqrt(pi) - cos(x)*erf(x) == (sqrt(pi)*exp(-1/4)*erf(x - I/2)/2 + sqrt(pi)*exp(-1/4)*erf(x + I/2)/2)/sq
rt(pi) - cos(x)*erf(x)
E        +  where 2*(sqrt(pi)*exp(-1/4)*erf(x - I/2)/4 + sqrt(pi)*exp(-1/4)*erf(x + I/2)/4)/sqrt(pi) - cos(x)*erf(x) = manualintegrate(sin(x)*erf(x), x)

sympy/integrals/tests/test_manual.py:30: AssertionError
====================================================================================== warnings summary =======================================================================================
venv/lib/python3.13/site-packages/_hypothesis_pytestplugin.py:459
  /home/salastro/.srcpkgs/sympy/venv/lib/python3.13/site-packages/_hypothesis_pytestplugin.py:459: UserWarning: Skipping collection of '.hypothesis' directory - this usually means you've expl
icitly set the `norecursedirs` pytest config option, replacing rather than extending the default ignores.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
                                                                                       DO *NOT* COMMIT!                                                                                       
=================================================================================== short test summary info ===================================================================================
FAILED sympy/integrals/tests/test_manual.py::test_manualintegrate_special - assert 2*(sqrt(pi)*exp(-1/4)*erf(x - I/2)/4 + sqrt(pi)*exp(-1/4)*erf(x + I/2)/4)/sqrt(pi) - cos(x)*erf(x) == (sqrt(
pi)*exp(-1/4)*erf(x - I/2)/2 + sqrt(pi)*exp(-1/4)*erf(x + I/2)/2)/sqrt(...
================================================================================ 1 failed, 1 warning in 10.16s ================================================================================
```

oscarbenjamin replied:
The `F` in the test needs to be of the same form that is returned by `manualintegrate`:
```python
In [11]: f
Out[11]: sin(x)⋅erf(x)

In [12]: f, F = sin(x)*erf(x), 2*(sqrt(pi)*exp(-Rational(1,4))*erf(x - I/2)/4 + sqrt(pi)*exp(-Rationa
       ⋮ l(1,4))*erf(x + I/2)/4)/sqrt(pi) - cos(x)*erf(x)

In [13]: f
Out[13]: sin(x)⋅erf(x)

In [14]: F
Out[14]: 
    -1/4    ⎛    ⅈ⎞       -1/4    ⎛    ⅈ⎞                
√π⋅ℯ    ⋅erf⎜x - ─⎟   √π⋅ℯ    ⋅erf⎜x + ─⎟                
            ⎝    2⎠               ⎝    2⎠                
─────────────────── + ───────────────────                
         2                     2                         
───────────────────────────────────────── - cos(x)⋅erf(x)
                   √π                                    

In [15]: F2 = integrate(f, manual=True)

In [16]: F2
Out[16]: 
  ⎛    -1/4    ⎛    ⅈ⎞       -1/4    ⎛    ⅈ⎞⎞                
  ⎜√π⋅ℯ    ⋅erf⎜x - ─⎟   √π⋅ℯ    ⋅erf⎜x + ─⎟⎟                
  ⎜            ⎝    2⎠               ⎝    2⎠⎟                
2⋅⎜─────────────────── + ───────────────────⎟                
  ⎝         4                     4         ⎠                
───────────────────────────────────────────── - cos(x)⋅erf(x)
                     √π                                      

In [17]: print(F2)
2*(sqrt(pi)*exp(-1/4)*erf(x - I/2)/4 + sqrt(pi)*exp(-1/4)*erf(x + I/2)/4)/sqrt(pi) - cos(x)*erf(x)
```

salastro committed:
Add tests for combinations of trig/hyperbolic with erf

salastro committed:
manualintegrate: add hyperbolic to LIATE heuristic

salastro committed:
manualintegrate: factor terms in results with erf*trig/hyperbolic products

salastro committed:
Enable integration test cases with erf*trig products
